ArmPkg/Drivers/SmmuDxe/IoMmu.c

```diff
@@ -190,6 +168,33 @@ UpdatePageTable (
     CurPhysicalAddress += EFI_PAGE_SIZE;
   }
 
+  // Invalidate TLBI Command.
+  // Current implementation of IoMmu protocol doesnt distinguish between stream id's or Smmu's.
+  // As a result we have to invalidate TLB for all SMMU's for the given virtual address to make
+  // sure that the correct smmu's streamid's page table's TLB is invalidated.
+  // To workaround the limitations of the IoMmu protocol, we globally set the page table root
+  // for all SMMU's and stream id's to be the same.
+  // Todo: Update the IoMmu protocol to support multiple SMMU's and stream id's so we don't have to
+  // invalidate all SMMU's TLB for a given virtual address, just the one that was updated.
+  if (!Valid) {
+    for (SmmuIndex = 0; SmmuIndex < mIoMmu->SmmuCount; SmmuIndex++) {
+      if (mIoMmu->SmmuInfo[SmmuIndex].Enabled) {
+        if (mIoMmu->SmmuInfo[SmmuIndex].RangeInvalidationSupported) {
+          Status = SmmuV3TLBInvalidateAddressRange (&mIoMmu->SmmuInfo[SmmuIndex], PhysicalAddress, EFI_SIZE_TO_PAGES (PhysicalAddressEnd - ALIGN_DOWN_BY (PhysicalAddress, EFI_PAGE_SIZE)));

```

Comment:
I would align down `PhysicalAddress` here, since the length more or less assumes its aligned down for it to not go over the next page.